File: core/core-web/src/test/java/ch/difty/scipamato/core/web/paper/PaperAttachmentProviderTest.java
package ch.difty.scipamato.core.web.paper;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.apache.wicket.model.IModel;
import org.apache.wicket.model.Model;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import ch.difty.scipamato.core.entity.PaperAttachment;

@ExtendWith(MockitoExtension.class)
class PaperAttachmentProviderTest {

    private PaperAttachmentProvider provider;

    @Mock
    private PaperAttachment mockAttachment1, mockAttachment2, mockAttachment3, mockAttachment4;

    private final List<PaperAttachment> attachments = new ArrayList<>();

    @BeforeEach
    void setUp() {
        attachments.addAll(Arrays.asList(mockAttachment1, mockAttachment2, mockAttachment3, mockAttachment4));
        provider = new PaperAttachmentProvider(Model.ofList(attachments));
    }

    @Test
    void degenerateConstruction_withNullSearchOrderModel1() {
        Assertions.assertThrows(NullPointerException.class, () -> new PaperAttachmentProvider(Model.ofList(null)));
    }

    @Test
    void providerSize_equals_conditionSize() {
        assertThat(provider.size()).isEqualTo(attachments.size());
    }

    @Test
    void iterator_fromStartWithPageSizeLargerThanActualSize_returnsAll() {
        assertThat(provider.iterator(0, 100))
            .toIterable()
            .containsExactly(mockAttachment1, mockAttachment2, mockAttachment3, mockAttachment4);
    }

    @Test
    void iterator_fromStartWithLimitingPageSize_returnsPageFullFromStart() {
        assertThat(provider.iterator(0, 2))
            .toIterable()
            .containsExactly(mockAttachment1, mockAttachment2);
    }

    @Test
    void iterator_fromIndex1WithLimitingPageSize_returnsPageFullFromIndex() {
        assertThat(provider.iterator(1, 2))
            .toIterable()
            .containsExactly(mockAttachment2, mockAttachment3);
    }

    @Test
    void gettingModel() {
        IModel<PaperAttachment> model = provider.model(mockAttachment1);
        assertThat(model.getObject()).isEqualTo(mockAttachment1);
    }
}

File: core/core-web/src/test/kotlin/ch/difty/scipamato/core/web/paper/PaperAttachmentProviderTest.kt
package ch.difty.scipamato.core.web.paper

import ch.difty.scipamato.core.entity.PaperAttachment
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldContainSame
import org.apache.wicket.model.Model
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class PaperAttachmentProviderTest {

    private lateinit var provider: PaperAttachmentProvider

    private val att1Dummy = PaperAttachment().apply { id = 1 }
    private val att2Dummy = PaperAttachment().apply { id = 2 }
    private val att3Dummy = PaperAttachment().apply { id = 3 }
    private val att4Dummy = PaperAttachment().apply { id = 4 }

    private val attachments = listOf(att1Dummy, att2Dummy, att3Dummy, att4Dummy)

    @BeforeEach
    fun setUp() {
        provider = PaperAttachmentProvider(Model.ofList(attachments))
    }

    @Test
    fun providerSize_equals_conditionSize() {
        provider.size() shouldBeEqualTo attachments.size.toLong()
    }

    @Test
    fun iterator_fromStartWithPageSizeLargerThanActualSize_returnsAll() {
        provider.iterator(0, 100).asSequence() shouldContainSame
            sequenceOf(att1Dummy, att2Dummy, att3Dummy, att4Dummy)
    }

    @Test
    fun iterator_fromStartWithLimitingPageSize_returnsPageFullFromStart() {
        provider.iterator(0, 2).asSequence() shouldContainSame sequenceOf(att1Dummy, att2Dummy)
    }

    @Test
    fun iterator_fromIndex1WithLimitingPageSize_returnsPageFullFromIndex() {
        provider.iterator(1, 2).asSequence() shouldContainSame sequenceOf(att2Dummy, att3Dummy)
    }

    @Test
    fun gettingModel() {
        val model = provider.model(att1Dummy)
        model.getObject() shouldBeEqualTo att1Dummy
    }
}
